### PR TITLE
Fixed many problems and added some features

### PR DESCRIPTION
**Pull request by TeddyNight**
I rewrote some code and had fixed many problems.

### CODE REVIEW

**Comment by TeddyNight** [comment]
Sorry about my English.I am a chinese.

File: README.md [hunk 1]
```diff
@@ -1,5 +1,9 @@
 # yosh - Your Own SHell
 
+This is fork from https://github.com/supasate/yosh
+
+And it's modifed by Ted,fixed some problems and made it more looks like a noraml shell
+
```
Comment (supasate) [inline README.md]: Don't need to specify this comment. I'll add detail and contributor into CHANGES.md file.

File: yosh/builtins/exit.py [hunk 2]
```diff
@@ -1,5 +1,4 @@
 from yosh.constants import *
 
-
 def exit(args):
```
Comment (supasate) [inline yosh/builtins/exit.py:4]: [PEP8] expected 2 blank lines, found 1

File: yosh/builtins/export.py [hunk 3]
```diff
@@ -0,0 +1,8 @@
+import os
+from yosh.constants import *
+
+
+def export(args):
+    var=args[0].split('=', 1 );
```
Comment (supasate) [inline yosh/builtins/export.py]: [PEP8]
- missing whitespace around operator
- whitespace before ')'
- statement ends with a semicolon

Need to fix it to `var = args[0].split('=', 1)`

**Comment by TeddyNight** [comment]
Thanks for commenting to me.I knew that I made too many changes on it.I thought it'll make many new problems.Let's find out them to fix. @supasate

File: yosh/shell.py [hunk 4]
```diff
@@ -9,59 +13,70 @@
...
+           # Tokenize the command input
+           cmd_tokens = tokenize(cmd)
+        except:
+           print "Error when receiving the command"
+        # Fix a bug with inputing nothing
+        try:
+           # Execute the command and retrieve new status
+           status = execute(cmd_tokens)
+        except OSError,e:
+              print e
```
Comment (supasate) [inline yosh/shell.py]: - Use `print(e)` for compatibility with Python 2 and 3.
- [PEP8] indentation is not a multiple of four

File: README.md (hunk 1 above)
Comment (TeddyNight) [inline README.md]: You can delete it in.I just write down something on my fork.Thanks.

File: yosh/shell.py [hunk 5]
```diff
@@ -9,59 +13,70 @@
...
+        try:
+           # Tokenize the command input
+           cmd_tokens = tokenize(cmd)
+        except:
+           print "Error when receiving the command"
+        # Fix a bug with inputing nothing
+        try:
+           # Execute the command and retrieve new status
+           status = execute(cmd_tokens)
+        except OSError,e:
```
Comment (supasate) [inline yosh/shell.py]: [PEP8] missing whitespace after ','

File: yosh/shell.py [hunk 6]
```diff
@@ -9,59 +13,70 @@
...
-        status = execute(cmd_tokens)
+        # Modifed by Ted,do not receive Ctrl signal
+        signal.signal(signal.SIGTSTP, signal.SIG_IGN)
+        signal.signal(signal.SIGINT,signal.SIG_IGN)  
+        #The bugs with receiving wrong command had fixed by Ted
+        try:
+           # Read command input
+           cmd = sys.stdin.readline()
+        except KeyboardInterrupt,e:
+           print e
```
Comment (supasate) [inline yosh/shell.py]: Use `print(e)` for compatibility with Python 2 and 3.

File: yosh/shell.py [hunk 7]
```diff
@@ -9,59 +13,70 @@
...
-            wpid, status = os.waitpid(pid, 0)
-
-            # Finish waiting if its child process exits normally or is
-            # terminated by a signal
-            if os.WIFEXITED(status) or os.WIFSIGNALED(status):
-                break
-
-    # Return status indicating to wait for next command in shell_loop
-    return SHELL_STATUS_RUN
+    global sh
```
Comment (supasate) [inline yosh/shell.py]: Don't use global.

File: yosh/shell.py [hunk 8]
```diff
@@ -9,59 +13,70 @@
...
+    signal.signal(signal.SIGINT,handler_kill)
+    # Written in beautiful sentences to run the command,modifed by Ted
```
Comment (supasate) [inline yosh/shell.py]: - Do not need to specify who is a modifier in code.
- Change comment to be more specific what it does. (Like `Spawn a child process`)

File: yosh/shell.py (hunk 7 above)
Comment (TeddyNight) [inline yosh/shell.py]: But this global is needed.Because of handler_kill,do you have any idea to instead of the global? @supasate

**Comment by supasate** [comment]
Hi @YuliangTan, first, thank you for your contribution! 

I'm trying to comment but the commit contains several PEP8 errors. 

Could you use [pep8](https://pypi.python.org/pypi/pep8) to check your code and fix according to its report?

**Comment by supasate** [comment]
@YuliangTan This is the result when I ran `pep8` to check.
image

**Comment by TeddyNight** [comment]
I just run it in the noraml.I don't use the tools to test my code before.I just test it by myself.Thanks for your advice.^_^

**Comment by TeddyNight** [comment]
I use the python 2.7.So I didn't think anything about running in the python 3 before.

**Comment by TeddyNight** [comment]
Now I fix the PEP8 error.Do you have any idea on my code?

**Comment by supasate** [comment]
Look much better! 

Why do you need `global sh` at [line 37](https://github.com/supasate/yosh/pull/3/files#diff-a7d9e02417dce0de832d21f61cf3d9e9R37)? Is it safe to remove?

One more idea is that the shell prompt including `full user@hostname /full/path/to/my/folder` is quite long when displaying. I think we can display only the last folder name of the current directory like in `zsh`.

For example, if we are in `/home/User/yosh`, just show `yosh $` would be suffice.

**Comment by TeddyNight** [comment]
I thought it needs global.But in fact,the global didn't need.How stupid I am.>_< Now the code is including your idea.^_^

File: yosh/shell.py [hunk 9]
```diff
@@ -54,14 +53,16 @@ def shell_loop():
         # Make it more looks like bash command prompt
         if os.getcwd() == os.getenv('HOME'):
             dir = "~"
+        elif os.getcwd() == os.getenv('HOME') + "/yosh":
+            dir = ""
```
Comment (supasate) [inline yosh/shell.py]: This case is not necessary. When we are in `/home/user/yosh`, it should display `yosh $` to reflect what directory that we are in.

File: yosh/shell.py [hunk 10]
```diff
@@ -54,14 +53,16 @@ def shell_loop():
...
         else:
-            dir = os.getcwd()
+            dir = os.getcwd().split('/')[-1]
         if os.geteuid() != 0:
             sys.stdout.write(
                              '[' + getpass.getuser() + '@' +
-                             socket.gethostname() + ' '+dir+']$ ')
+                             socket.gethostname().split('.')[0] + ' '+dir+']$ ')
```
Comment (supasate) [inline yosh/shell.py]: Add spaces before and after `+` sign.

File: yosh/shell.py [hunk 11]
```diff
@@ -54,14 +53,16 @@ def shell_loop():
...
         else:
-            sys.stdout.write('[root@'+socket.gethostname()+' '+dir+']# ')
+            sys.stdout.write('[root@'+socket.gethostname().split('.')[0]+' '+dir+']# ')
```
Comment (supasate) [inline yosh/shell.py]: Add spaces before and after `+` sign.

**Comment by supasate** [comment]
I add some more comments in the latest commit. Everything else looks fine.

File: yosh/shell.py (hunk 9 above)
Comment (TeddyNight) [inline yosh/shell.py]: But I think the /home/user/yosh is not the real home directory.

**Comment by TeddyNight** [comment]
OK,I did it.But I think /home/user/yosh shouldn't display yosh $ .

**Comment by supasate** [comment]
Why do you think it shouldn't display `yosh $` on `/home/user/yosh`?

IMO, `/home/user/yosh` is just a folder and I have no reason to treat it more special than `/home/user/my_other_folder`.

**Comment by TeddyNight** [comment]
I don't think yosh $ on /home/user/yosh is a good choice.As we know,$HOME in the noraml shell is /home/user/.And it won't be /home/user/my_other_folder.

**Comment by supasate** [comment]
For example, let assume our terminal uses our shell as its default shell.

If we start, we are at `/home/user/`, the prompt shows 

`[user@host ~]$`.

Then, we create 2 folders:

``` bash
mkdir test
mkdir yosh
```

If we go into test folder with `cd test`, the prompt will show 

`[user@host test]$`.

However, if we change the directory into `/home/user/yosh` by `cd ../yosh`, the prompt will not show any folder 

`[user@host ]$`. 

I think this will confuse users why our shell does not show their current directory.

Let's take a look at Z shell (`zsh`).  

image

This will let users know they are in the folder name `zsh` even it is the same as the shell name.

What do you think?

**Comment by supasate** [comment]
One more reason is that this project can be forked and renamed into anything. So, we shouldn't hard code the project name into the code.

**Comment by TeddyNight** [comment]
Now I cause the /home/user/yosh.I write a setup.py file.That mean the user can run the binary.Just    enter "yosh".

File: setup.py [hunk 12]
```diff
@@ -0,0 +1,16 @@
+#!/usr/bin/env python
+
+from setuptools import setup
+
+setup(name='yosh',
+      version='1.0',
+      description='Your own shell',
+      author='Supasate',
+      author_email='tanyuliang2@gmail.com',
```
Comment (supasate) [inline setup.py:9]: supasate.c@gmail.com

File: setup.py [hunk 13]
```diff
@@ -0,0 +1,16 @@
+#!/usr/bin/env python
+
+from setuptools import setup
+
+setup(name='yosh',
+      version='1.0',
+      description='Your own shell',
+      author='Supasate',
```
Comment (supasate) [inline setup.py:8]: Supasate Choochaisri

**Comment by supasate** [comment]
Looks cool with setup.py.

Fix name and email in my latest comments and I think we should stop at this point for this PR. I'll merge it and tag it as a new version.

Another change can be pushed as a new PR. Thanks!!

**Pull request merged**